lisa/tools/dmesg.py

```diff
@@ -0,0 +1,72 @@
+import re
+from typing import Any, List, Optional
+
+from lisa.executable import Tool
+from lisa.util import LisaException
+from lisa.util.process import ExecutableResult
+
+
+class Dmesg(Tool):
+    # meet any pattern will be considered as potential panic line.
+    __panic_patterns = [
+        re.compile(r"^[^[].+$"),
+        re.compile("Call Trace"),
+        re.compile("rcu_sched self-detected stall on CPU"),
+        re.compile("rcu_sched detected stalls on"),
+        re.compile("BUG: soft lockup"),
+    ]
+    # ignore some return lines, which shouldn't be a panic line.
+    __panic_ignorable_patterns = [
+        re.compile(r"this clock source is slow\. Consider trying other clock sources"),
+        re.compile(r"If you want to keep using the local clock\, then add\:"),
+        re.compile('"trace_clock=local"'),
+        re.compile(r"on the kernel command line"),
+    ]
+
+    @property
+    def command(self) -> str:
+        return "dmesg"
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        self._cached_result: Optional[ExecutableResult] = None
+
+    def _check_exists(self) -> bool:
+        return True
+
+    def check_kernel_panic(
+        self, force_run: bool = False, throw_error: bool = True
+    ) -> str:
+        command_output = self._run(force_run=force_run)
+        if command_output.exit_code != 0:
+            raise LisaException(f"exit code should be zero: {command_output.exit_code}")
+        panic_lines: List[str] = []
+        for line in command_output.stdout.splitlines(keepends=False):
+            for pattern in self.__panic_patterns:
+                if pattern.search(line):
+                    is_panic_matched = True
+                    for ignorable_pattern in self.__panic_ignorable_patterns:
+                        if ignorable_pattern.search(line):
+                            is_panic_matched = False
+                            break
+                    if is_panic_matched:
+                        panic_lines.append(line)
+                        # match one rule, so skip for other patterns
+                        break
```

Comment:
No need to change this (this is clear enough!), but here's a thing that Python has that's pretty cool and can be used in situations like this:

```suggestion
                    for ignorable_pattern in self.__panic_ignorable_patterns:
                        if ignorable_pattern.match(line):
                            break
                    else:
                        panic_lines.append(line)
                        self._log.debug(f"found error line: {line}")
                        # match one rule, so skip for other patterns
                        break
```